### PR TITLE
Don't use symex_dynamic:: magic string in multiple places

### PR DESCRIPTION
**Pull request by thk123**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/util/simplify_expr_pointer.cpp [hunk 1]
```diff
@@ -9,6 +9,7 @@ Author: Daniel Kroening, kroening@kroening.com
 #include "simplify_expr_class.h"
 
 #include <cassert>
+#include <goto-symex/goto_symex.h>
```
Comment (smowton) [inline src/util/simplify_expr_pointer.cpp]: I suspect that's a bad cross-module dependency (but then, so is utils knowing about symex internal magic)

Comment (thk123) [reply src/util/simplify_expr_pointer.cpp]: Yeah looks like this particular simplification shouldn't be in util. Happy to either remove this particular usage - or pull the macro into `util`?

**Review by tautschnig — CHANGES_REQUESTED** [review]
I'd suggest `util/pointer_predicates.h` as a place to store the macro and thus avoid the dependency problem noted by @smowton. There is a bigger story here with f1ff824d00fb4 from #2646 doing essential cleanup.

**Comment by thk123** [comment]
@tautschnig @smowton dependency comment [addressed](https://github.com/diffblue/cbmc/compare/145bd2d2936cc766b3901203fe3fba283e108830..091e987f735a5c6fa216f2fac4335666328a403a)

**Review by smowton — APPROVED** [review]
Fine with this for now, but please make an issue to avoid non-symex modules having to know about a symex-specific magic identifier

**Comment by thk123** [comment]
Done in #3526

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: abe2102d3950a2a047664c0570f60346694803c9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/93531642
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Comment by thk123** [comment]
@tautschnig this is ready for a final review

**Review by tautschnig — APPROVED** [review]
My apologies for delaying this for such a long time. Looks good to me, but needs a rebase.

**Comment by thk123** [comment]
Rebased and found another one.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 042ad7477035c0f50705086b0176da860868f9b7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102308171

**Comment by thk123** [comment]
Time to play wack-a-mole with code owners - I need a goto-instrument code owner review:
@martin-cs @chrisr-diffblue @peterschrammel @danpoe @hannes-steffenhagen-diffblue

**Review by chrisr-diffblue — APPROVED** [review]

**Pull request merged**